Indexed configuration properties not injected by @ConfigProperty

## Describe the bug
When using indexed properties, like described in the [docs](https://quarkus.io/guides/config-reference#indexed-properties), I can only access a collection programatically, but not via injection with `@ConfigProperty`.

## To Reproduce

```java
@ApplicationScoped
public class MyBean {

    @ConfigProperty(name = "complex.list")
    List<String> complexList;

    public void method() {
        List<String> values = ConfigProvider.getConfig().getValues("complex.list", String.class);
        System.out.println(values);
    }
}
```

and properties:

```
complex.list[0]=123
complex.list[1]=234
complex.list[2]=456
```

Fails with:

```
Caused by: java.lang.RuntimeException: Failed to start quarkus
  [...]
Caused by: javax.enterprise.inject.spi.DeploymentException: Failed to load config value of type class java.lang.String for: complex.list
	at io.quarkus.arc.runtime.ConfigRecorder.validateConfigProperties(ConfigRecorder.java:42)
	at io.quarkus.deployment.steps.ConfigBuildStep$validateConfigProperties1249763973.deploy_0(ConfigBuildStep$validateConfigProperties1249763973.zig:332)
	at io.quarkus.deployment.steps.ConfigBuildStep$validateConfigProperties1249763973.deploy(ConfigBuildStep$validateConfigProperties1249763973.zig:40)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:713)
	... 12 more
Caused by: java.util.NoSuchElementException: SRCFG00014: The config property complex.list is required but it could not be found in any config source
	at io.smallrye.config.SmallRyeConfig.convertValue(SmallRyeConfig.java:257)
	at io.smallrye.config.inject.ConfigProducerUtil.getValue(ConfigProducerUtil.java:84)
	at io.quarkus.arc.runtime.ConfigRecorder.validateConfigProperties(ConfigRecorder.java:39)
	... 15 more
```

However, if I uncomment the `complexList`, the programmatic lookup, like described in the docs works. But, I'd expect the same to work via injection.

I'm using Quarkus `2.0.0.Final` with JDK 16

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ConfigBuildStep.java
@@ -15,6 +15,7 @@ import static org.jboss.jandex.AnnotationTarget.Kind.METHOD_PARAMETER;
 import static org.jboss.jandex.AnnotationValue.createStringValue;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
@@ -182,9 +183,18 @@ public class ConfigBuildStep {
 
         Set<ConfigValidationMetadata> propertiesToValidate = new HashSet<>();
         for (ConfigPropertyBuildItem configProperty : configProperties) {
+            String rawTypeName = configProperty.getPropertyType().name().toString();
+            List<String> actualTypeArgumentNames = Collections.emptyList();
+            if (configProperty.getPropertyType().kind() == Kind.PARAMETERIZED_TYPE) {
+                List<Type> argumentTypes = configProperty.getPropertyType().asParameterizedType().arguments();
+                actualTypeArgumentNames = new ArrayList<>(argumentTypes.size());
+                for (Type argumentType : argumentTypes) {
+                    actualTypeArgumentNames.add(argumentType.name().toString());
+                }
+
+            }
             propertiesToValidate.add(new ConfigValidationMetadata(configProperty.getPropertyName(),
-                    configProperty.getPropertyType().name().toString(),
-                    configProperty.getDefaultValue()));
+                    rawTypeName, actualTypeArgumentNames, configProperty.getDefaultValue()));
         }
 
         recorder.validateConfigProperties(propertiesToValidate);

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/config/IndexedPropertiesInjectionTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/config/IndexedPropertiesInjectionTest.java
@@ -15,7 +15,6 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import javax.enterprise.context.ApplicationScoped;
-import javax.enterprise.inject.spi.DeploymentException;
 import javax.inject.Inject;
 
 import org.eclipse.microprofile.config.inject.ConfigProperty;
@@ -48,8 +47,7 @@ public class IndexedPropertiesInjectionTest {
                                     "optionals.indexed[1]=b\n" +
                                     "supplier.indexed[0]=a\n" +
                                     "supplier.indexed[1]=b\n"),
-                            "application.properties"))
-            .setExpectedException(DeploymentException.class);
+                            "application.properties"));
 
     @Inject
     IndexedBean indexedBean;

--- a/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigRecorder.java
+++ b/extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/ConfigRecorder.java
@@ -1,5 +1,7 @@
 package io.quarkus.arc.runtime;
 
+import java.lang.reflect.Type;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
@@ -9,6 +11,8 @@ import javax.enterprise.inject.spi.DeploymentException;
 import org.eclipse.microprofile.config.Config;
 import org.eclipse.microprofile.config.ConfigProvider;
 
+import io.quarkus.arc.impl.ParameterizedTypeImpl;
+import io.quarkus.runtime.annotations.RecordableConstructor;
 import io.quarkus.runtime.annotations.Recorder;
 import io.smallrye.config.ConfigMappings;
 import io.smallrye.config.ConfigMappings.ConfigClassWithPrefix;
@@ -30,17 +34,27 @@ public class ConfigRecorder {
         }
 
         for (ConfigValidationMetadata property : properties) {
-            Class<?> propertyClass = load(property.getType(), cl);
+            Class<?> propertyType = load(property.getRawTypeName(), cl);
+            Type effectivePropertyType = propertyType;
             // For parameterized types and arrays, we only check if the property config exists without trying to convert it
-            if (propertyClass.isArray() || (propertyClass.getTypeParameters().length > 0 && propertyClass != Map.class)) {
-                propertyClass = String.class;
+            if (propertyType.isArray() || (propertyType.getTypeParameters().length > 0 && propertyType != Map.class
+                    && propertyType != List.class && propertyType != Set.class)) {
+                effectivePropertyType = String.class;
+            } else if (property.getActualTypeArgumentNames().size() > 0) {
+                // this is a really simplified way of constructing the generic types, but we don't need anything more complex
+                // here due to what SR Config checks (which is basically if the type is a collection)
+                Type[] genericTypes = new Type[(property.getActualTypeArgumentNames().size())];
+                for (int i = 0; i < property.getActualTypeArgumentNames().size(); i++) {
+                    genericTypes[i] = load(property.getActualTypeArgumentNames().get(i), cl);
+                }
+                effectivePropertyType = new ParameterizedTypeImpl(propertyType, genericTypes);
             }
 
             try {
-                ConfigProducerUtil.getValue(property.getName(), propertyClass, property.getDefaultValue(), config);
+                ConfigProducerUtil.getValue(property.getName(), effectivePropertyType, property.getDefaultValue(), config);
             } catch (Exception e) {
                 throw new DeploymentException(
-                        "Failed to load config value of type " + propertyClass + " for: " + property.getName(), e);
+                        "Failed to load config value of type " + effectivePropertyType + " for: " + property.getName(), e);
             }
         }
     }
@@ -94,15 +108,16 @@ public class ConfigRecorder {
 
     public static class ConfigValidationMetadata {
         private String name;
-        private String type;
+        private String rawTypeName;
+        private List<String> actualTypeArgumentNames;
         private String defaultValue;
 
-        public ConfigValidationMetadata() {
-        }
-
-        public ConfigValidationMetadata(final String name, final String type, final String defaultValue) {
+        @RecordableConstructor
+        public ConfigValidationMetadata(final String name, final String rawTypeName, List<String> actualTypeArgumentNames,
+                final String defaultValue) {
             this.name = name;
-            this.type = type;
+            this.rawTypeName = rawTypeName;
+            this.actualTypeArgumentNames = actualTypeArgumentNames;
             this.defaultValue = defaultValue;
         }
 
@@ -114,39 +129,33 @@ public class ConfigRecorder {
             this.name = name;
         }
 
-        public String getType() {
-            return type;
+        public String getRawTypeName() {
+            return rawTypeName;
         }
 
-        public void setType(final String type) {
-            this.type = type;
+        public List<String> getActualTypeArgumentNames() {
+            return actualTypeArgumentNames;
         }
 
         public String getDefaultValue() {
             return defaultValue;
         }
 
-        public void setDefaultValue(final String defaultValue) {
-            this.defaultValue = defaultValue;
-        }
-
         @Override
-        public boolean equals(final Object o) {
-            if (this == o) {
+        public boolean equals(Object o) {
+            if (this == o)
                 return true;
-            }
-            if (o == null || getClass() != o.getClass()) {
+            if (o == null || getClass() != o.getClass())
                 return false;
-            }
-            final ConfigValidationMetadata that = (ConfigValidationMetadata) o;
-            return name.equals(that.name) &&
-                    type.equals(that.type) &&
-                    Objects.equals(defaultValue, that.defaultValue);
+            ConfigValidationMetadata that = (ConfigValidationMetadata) o;
+            return name.equals(that.name) && rawTypeName.equals(that.rawTypeName)
+                    && actualTypeArgumentNames.equals(that.actualTypeArgumentNames)
+                    && Objects.equals(defaultValue, that.defaultValue);
         }
 
         @Override
         public int hashCode() {
-            return Objects.hash(name, type, defaultValue);
+            return Objects.hash(name, rawTypeName, actualTypeArgumentNames, defaultValue);
         }
     }
 }

